### PR TITLE
feat: :alembic: move fast and break things

### PR DESCRIPTION
**Pull request by pyaillet**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -9,3 +9,4 @@ edition = "2018"
 [dependencies]
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
+ureq = "1.5.1"
```
Comment (charlyx) [inline Cargo.toml]: I didn't know ureq!
I believe blocking is fine for now. 
What was wrong with reqwest?

File: src/oci/image/config.rs [hunk 2]
```diff
@@ -71,13 +71,13 @@ pub struct Config {
     history: Vec<History>,
 }
 
-pub fn parse_config(content: &str) -> Result<Config, Box<dyn error::Error>> {
+pub fn from_str(content: &str) -> Result<Config, Box<dyn error::Error>> {
```
Comment (charlyx) [inline src/oci/image/config.rs]: Why aren't you implementing FromStr?

https://doc.rust-lang.org/std/str/trait.FromStr.html

File: src/oci/image/mod.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    fn tag(self) -> Result<String, Box<dyn error::Error>>;
+
+    fn digest(self) -> Result<String, Box<dyn error::Error>>;
+}
+
+#[derive(Debug, PartialEq)]
+pub struct ImageReference(String);
+
+impl Reference for ImageReference {
+    fn registry(self) -> Result<String, Box<dyn error::Error>> {
```
Comment (charlyx) [inline src/oci/image/mod.rs]: I can't see any error. 
Shouldn't it be an Option instead though?

File: src/oci/image/mod.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        );
+    }
+
+    #[test]
+    fn registry_ok() {
+        let image = ImageReference("localhost:5000/myorg/myimage/test:local".to_string());
+
+        let r = image.registry();
+
+        assert!(r.is_ok(), "Manifest parsing failed: `{}`", r.err().unwrap());
```
Comment (charlyx) [inline src/oci/image/mod.rs]: Manifest?

File: src/oci/registry.rs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    let res = ureq::get("http://httpbin.org/get").call().into_string()?;
+
+    manifest::from_str(&res)
+}
+
+#[test]
+fn get_ok() {
+    let r = get("localhost:5000/test:local".to_string());
+
+    if r.is_err() {
```
Comment (charlyx) [inline src/oci/registry.rs]: Why not using assert as in previous tests?

File: src/oci/image/mod.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    fn registry(self) -> String {
+        match self.0.find("/") {
+            Some(x) => self.0[..x].to_string(),
+            None => "docker.io".to_string(),
+        }
+    }
+
+    fn fullname(self) -> String {
+        match self.0.find("/") {
+            Some(x) => self.0[x + 1..].to_string(),
```
Comment (charlyx) [inline src/oci/image/mod.rs:30]: Can't you use "into" instead?

File: Cargo.toml (hunk 1 above)
Comment (pyaillet) [reply Cargo.toml]: I didn't find how to use it without async... 🤷‍♂️

File: src/oci/image/config.rs (hunk 2 above)
Comment (pyaillet) [reply src/oci/image/config.rs]: In fact, it's already there from the derivation, getting rid of this useless method.

File: src/oci/image/mod.rs (hunk 3 above)
Comment (pyaillet) [reply src/oci/image/mod.rs]: You're right, it's useless.

File: src/oci/image/mod.rs (hunk 4 above)
Comment (pyaillet) [reply src/oci/image/mod.rs]: Got rid of this

File: src/oci/image/config.rs (hunk 2 above)
Comment (charlyx) [reply src/oci/image/config.rs]: Good, that's what I thought.

File: Cargo.toml (hunk 1 above)
Comment (charlyx) [reply Cargo.toml]: reqwest::blocking::get? 

There is an example in the README: https://github.com/seanmonstar/reqwest

Is it not working as expected?

File: src/oci/registry.rs [hunk 7]
```diff
@@ -0,0 +1,24 @@
+use crate::oci::image::manifest::Manifest;
+use std::error;
+
+pub fn get_manifest(path: String) -> Result<Manifest, Box<dyn error::Error>> {
+    let res = ureq::get("http://httpbin.org/get").call().into_string()?;
+    let m: Result<Manifest, serde_json::Error> = serde_json::from_str(&res);
+
+    match m {
+        Ok(m) => Ok(m),
+        Err(e) => Err(Box::new(e)),
```
Comment (charlyx) [inline src/oci/registry.rs]: I believe you can get rid of the match by calling map_err on the match. 
Do you have to box the error?

Comment (charlyx) [reply src/oci/registry.rs]: http://www.sheshbabu.com/posts/rust-error-handling/

Comment (pyaillet) [reply src/oci/registry.rs]: I was not happy with this and found something even better.
Give it a look ! ;)

File: src/oci/image/mod.rs (hunk 6 above)
Comment (pyaillet) [reply src/oci/image/mod.rs:30]: I'm not sure to understand the difference, might read things on it before.

File: src/oci/registry.rs (hunk 5 above)
Comment (pyaillet) [reply src/oci/registry.rs]: debugging purpose ;)

File: src/oci/image/mod.rs (hunk 6 above)
Comment (pyaillet) [reply src/oci/image/mod.rs:30]: From what I read [here](https://users.rust-lang.org/t/what-is-the-idiomatic-way-to-convert-str-to-string/12160/8) `into` would work but would be preferred when converting from another type.
Not sure what to choose though...

File: Cargo.toml (hunk 1 above)
Comment (pyaillet) [reply Cargo.toml]: Switched back to `reqwest`

File: src/oci/image/config.rs [hunk 8]
```diff
@@ -132,7 +124,7 @@ fn test_parse_config_ok() {
             ]
         }"#;
 
-    let m = parse_config(c);
+    let m: Result<Config, serde_json::Error> = serde_json::from_str(c);
```
Comment (charlyx) [inline src/oci/image/config.rs]: I believe the test function could return this result type so we could use the question mark operator and do some proper check on config?

Comment (pyaillet) [reply src/oci/image/config.rs]: Indeed, I just updated it.

**Pull request merged**